lib/nostrum/util.ex

```diff
@@ -145,24 +145,29 @@ defmodule Nostrum.Util do
   @spec gateway() :: String.t()
   def gateway do
     case :ets.lookup(:gateway_url, "url") do
-      [] -> get_new_gateway_url()
-      [{"url", url, shards}] -> {url, shards}
+      [{"url", url, shards}] ->
+        {url, shards}
+      [] -> 
+        case get_new_gateway_url() do
+          {:ok, url, shards} -> {url, shards}
+          {:error, message} -> raise(message)
+        end
     end
   end
 
   defp get_new_gateway_url do
     case Api.request(:get, Constants.gateway_bot(), "") do
-      {:error, %{status_code: code, message: message}} ->
-        raise(Nostrum.Error.ApiError, status_code: code, message: message)
-
       {:ok, body} ->
         body = Poison.decode!(body)
 
         url = body["url"]
         shards = if body["shards"], do: body["shards"], else: 1
 
         :ets.insert(:gateway_url, {"url", url, shards})
-        {url, shards}
+
+        {:ok, url, shards}
+      
+      {:error, message} -> {:error, message}
```

Comment:
I'm not a super big fan of a case statement like this. Rather I'd be more inclined to fix the case statement we had before to handle an error struct instead of the map. 

Your approach is a little more clear so if you want to go that way maybe we can find a middle ground or just go with your method. Let me know your thoughts on the matter!

Reply:
Are you talking about the case statement in `get_new_gateway_url`? What specifically are you not a fan of?

I waffled on the nested case statement in `gateway`, trying out some with statements, but I couldn't quite express multiple steps in a failure path the same way that I could in the success path. Maybe I just have to forget "success" and "failure" and invert the two...

Reply:
Yeah, I was talking about the case statement in `get_new_gateway_url` that has a case that looks like `{:error, message} -> {:error, message}`. In my opinion it seems kind of silly to handle the case like this just to pass it up and raise an error one step higher.

The intended outcome of this function is to return the number of shards and gateway url, and the user will never interact with this function unless we reach the case where their token is invalid and we throw an error. In this case I think it's ok to throw the error at the lowest level. We can just match on the `Nostrum.Error.ApiError`

Let me know if you feel otherwise and we can investigate some other way to handle this!

Reply:
Gotcha. That makes sense to me.